File: Ant/chapter2/python/zero_one_knapsack_problem2.py
MAX_N = 100
MAX_V = 100
INF = float('inf')
V = MAX_N * MAX_V


def ns(f):
    return next(f).strip()


with open("../testset/zero_one_knapsack_problem/test1.txt", 'r') as f:
    N, W = map(int, ns(f).split())
    P = [list(map(int, ns(f).split())) for _ in range(N)]   # (w, v)


dp = [[INF] * (V + 1) for _ in range(N + 1)]
dp[0][0] = 0
for i in range(N):
    for j in range(V + 1):
        if j < P[i][1]:
            dp[i + 1][j] = dp[i][j]
        else:
            dp[i + 1][j] = min(dp[i][j], dp[i][j - P[i][1]] + P[i][0])
res = 0
for j in range(V + 1):
    if dp[N][j] <= W:
        res = j
print(res)

File: Ant/chapter2/cpp/zero_one_knapsack_problem2.cpp
#include <algorithm>
#include <fstream>
#include <stdio.h>

#define MAX_N 100
#define MAX_V 100
#define INF 2'000'000'000

using namespace std;

int N, W;
int w[MAX_N], v[MAX_N];
int dp[MAX_N + 1][MAX_N * MAX_V + 1];

int main() {
  ifstream ifs("../testset/zero_one_knapsack_problem/test1.txt");
  ifs >> N >> W;
  for (int i = 0; i < N; ++i) {
    ifs >> w[i] >> v[i];
  }
  int V = MAX_N * MAX_V;
  fill(dp[0], dp[0] + V + 1, INF);
  dp[0][0] = 0;
  for (int i = 0; i < N; ++i) {
    for (int j = 0; j <= V; ++j) {
      if (j < v[i]) {
        dp[i + 1][j] = dp[i][j];
      } else {
        dp[i + 1][j] = min(dp[i][j], dp[i][j - v[i]] + w[i]);
      }
    }
  }
  int res = 0;
  for (int j = 0; j <= V; ++j) {
    if (dp[N][j] <= W) {
      res = j;
    }
  }
  printf("%d\n", res);
}